include/psa/crypto_values.h

```diff
@@ -1255,6 +1278,79 @@
 #define PSA_ALG_IS_RANDOMIZED_ECDSA(alg)                                \
     (PSA_ALG_IS_ECDSA(alg) && !PSA_ALG_ECDSA_IS_DETERMINISTIC(alg))
 
+/** Edwards-curve digital signature algorithm without prehashing (PureEdDSA),
+ * using standard parameters.
+ *
+ * Contexts are not supported in the current version of this specification
+ * because there is no suitable signature interface that can take the
+ * context as a parameter. A future version of this specification may add
+ * suitable functions and extend this algorithm to support contexts.
+ *
+ * PureEdDSA requires an elliptic curve key on a twisted Edwards curve.
+ * In this specification, the following curves are supported:
+ * - #PSA_ECC_FAMILY_TWISTED_EDWARDS, 255-bit: Ed25519 as specified
+ *   in RFC 8032.
+ *   The curve is Edwards25519.
+ *   The hash function used internally is SHA-512.
+ * - #PSA_ECC_FAMILY_TWISTED_EDWARDS, 448-bit: Ed448 as specified
```

Comment:
Is 448-bit consistent with the way we define private key sizes, given that the private key is a 57-byte string?

Reply:
Yes, it is. If we go with @stevew817 's suggestion and use 255 for Ed25519, then we are signalling the bitlength of the prime defining the underlying field and not some key length.

Reply:
The PSA Crypto API spec does describe the [exported] private key size under `psa_export_key()` as the format for exported key pairs is the private key.

For Weierstrass curves it ties the output size to the key bit size:

> For Weierstrass curve families `PSA_ECC_FAMILY_SECT_XX`, `PSA_ECC_FAMILY_SECP_XX`, `PSA_ECC_FAMILY_FRP` and `PSA_ECC_FAMILY_BRAINPOOL_P_R1`, the content of the `privateKey` field of the `ECPrivateKey` format defined by Elliptic Curve Private Key Structure [RFC5915].
>
> This is a `ceiling(m/8)`-byte string in big-endian order where m is the key size in bits.

For Montgomery curves it does provide the computed value, but is explicit about the byte size of the exported key:

> For curve family `PSA_ECC_FAMILY_MONTGOMERY`, the scalar value of the ‘private key’ in little-endian order as defined by Elliptic Curves for Security [RFC7748] §6. The value must have the forced bits set to zero or one as specified by `decodeScalar25519()` and `decodeScalar448()` in [RFC7748] §5.
>
> This is a `ceiling(m/8)`-byte string where m is the key size in bits. This is 32 bytes for Curve25519, and 56 bytes for Curve448.

For the Edwards curves we could drop the computational definition and just provide the byte-size of the exported keys.

Reply:
In addition to [`psa_export_key`](https://armmbed.github.io/mbed-crypto/html/api/keys/management.html?highlight=psa_export_key#c.psa_export_key), there's information about what is considered the “size” of an elliptic curve key in the specification of [ECDH](https://armmbed.github.io/mbed-crypto/html/api/ops/ka.html?highlight=psa_alg_ecdh#c.PSA_ALG_ECDH) (probably not the right place for that!):

> The shared secret produced by key agreement is the x-coordinate of the shared secret point. It is always ceiling(m / 8) bytes long where m is the bit size associated with the curve, i.e. the bit size of the order of the curve’s coordinate field. When m is not a multiple of 8, the byte containing the most significant bit of the shared secret is padded with zero bits. The byte order is either little-endian or big-endian depending on the curve type.
>
> * For Montgomery curves (curve family PSA_ECC_FAMILY_MONTGOMERY), the shared secret is the x-coordinate of Z = d_A Q_B = d_B Q_A in little-endian byte order.
>
>     * For Curve25519, this is the X25519 function defined in Curve25519: new Diffie-Hellman speed records [Curve25519]. The bit size m is 255.
>     * For Curve448, this is the X448 function defined in Ed448-Goldilocks, a new elliptic curve [Curve448]. The bit size m is 448.
> * For Weierstrass curves (curve families PSA_ECC_FAMILY_SECP_XX, PSA_ECC_FAMILY_SECT_XX, PSA_ECC_FAMILY_BRAINPOOL_P_R1 and PSA_ECC_FAMILY_FRP) the shared secret is the x-coordinate of Z = h d_A Q_B = h d_B Q_A in big-endian byte order. This is the Elliptic Curve Cryptography Cofactor Diffie-Hellman primitive defined by SEC 1: Elliptic Curve Cryptography [SEC1] §3.3.2 as, and also as ECC CDH by NIST Special Publication 800-56A: Recommendation for Pair-Wise Key-Establishment Schemes Using Discrete Logarithm Cryptography [SP800-56A] §5.7.1.2.
>
>     * Over prime fields (curve families PSA_ECC_FAMILY_SECP_XX, PSA_ECC_FAMILY_BRAINPOOL_P_R1 and PSA_ECC_FAMILY_FRP), the bit size is m = ceiling(log_2(p)) for the field F_p.
>
>     * Over binary fields (curve families PSA_ECC_FAMILY_SECT_XX), the bit size is m for the field F_{2^m}.

Let's try to find a better compromise between consistency, correctness, comprehensibility and backward compatibility.

Reply:
I think [Steven's comment on 255](https://github.com/ARMmbed/mbedtls/pull/4174/commits/2f844a69596461eac0ce6e278145d3604c222ffe#r582899219) also applies here: in the Montgomery family we use 448, so let's be consistent and use 448 here too.

(Also, as Janos said, let's be nice to users: 448 is what's in the common name, require people to use another number would be confusing.)

Reply:
@yanesca - we are already inconsistent with a "what's in the common name" formula - secp224k1 is defined to use `key_bits = 225`.

However, [I think] we are currently consistent with the rule "for ECC keys, `key_bits` is the bit size of the order of the curve’s coordinate field". I.e. for prime fields over `p`, it is `ceiling(log2(p))`. (secp224k1 uses a 225 bit prime number.)

If we use the latter rule, then Ed25519 would have `key_bits = 255` and Ed448 has `key_bits = 448`.

The ECDH description can be simplified by removing some of the formulae, as the normative references provide the detail of the algorithm implementation. Comparing the description with that for `psa_export_public_key()`, we can make the key bit-size -> output relationship specific to the curve type, instead of trying to provide a one size fits all definition for the shared secret size.

Reply:
For the various reasons exposed in this thread, I've kept 448.